File: scripts/fetch-valuations.mjs
#!/usr/bin/env node
/**
 * 估值数据拉取脚本
 * 
 * 使用 Financial Modeling Prep (FMP) API 拉取公司估值数据，
 * 输出符合应用导入格式的 JSON 文件。
 *
 * 用法:
 *   node scripts/fetch-valuations.mjs AAPL MSFT GOOGL
 *   node scripts/fetch-valuations.mjs AAPL --key YOUR_FMP_KEY
 *   node scripts/fetch-valuations.mjs --tickers AAPL,MSFT,GOOGL
 *
 * API Key 配置（优先级从高到低）:
 *   1. 命令行参数 --key YOUR_KEY
 *   2. 环境变量 FMP_API_KEY
 *   3. 项目根目录 .env 文件中的 FMP_API_KEY=xxx
 *
 * 免费申请 FMP API Key: https://financialmodelingprep.com/developer/docs/
 * 免费额度: 250次/天
 *
 * 输出:
 *   ./valuation-import-YYYY-MM-DD.json（可直接粘贴到应用的导入框）
 */

import { readFileSync, writeFileSync, existsSync } from 'fs';
import { resolve, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(__dirname, '..');

// ── 解析参数 ─────────────────────────────────────────────────────────────

function parseArgs() {
  const args = process.argv.slice(2);
  let key = null;
  const tickers = [];

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--key' && args[i + 1]) {
      key = args[++i];
    } else if (args[i] === '--tickers' && args[i + 1]) {
      tickers.push(...args[++i].split(',').map(t => t.trim().toUpperCase()));
    } else if (!args[i].startsWith('--')) {
      tickers.push(args[i].toUpperCase());
    }
  }

  return { key, tickers };
}

// ── 读取 API Key ─────────────────────────────────────────────────────────

function getApiKey(cliKey) {
  if (cliKey) return cliKey;
  if (process.env.FMP_API_KEY) return process.env.FMP_API_KEY;

  // 尝试读取 .env 文件
  const envPath = resolve(ROOT, '.env');
  if (existsSync(envPath)) {
    const content = readFileSync(envPath, 'utf-8');
    const match = content.match(/^FMP_API_KEY\s*=\s*(.+)$/m);
    if (match) return match[1].trim();
  }
  return null;
}

// ── FMP API 请求 ─────────────────────────────────────────────────────────

const BASE = 'https://financialmodelingprep.com/api/v3';

async function fmpGet(path, apiKey) {
  const sep = path.includes('?') ? '&' : '?';
  const url = `${BASE}${path}${sep}apikey=${apiKey}`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`FMP API error ${res.status}: ${path}`);
  return res.json();
}

// ── 拉取单只股票数据 ─────────────────────────────────────────────────────

async function fetchTicker(ticker, apiKey) {
  console.log(`  ⏳ ${ticker} — 拉取中...`);

  const [profile, keyMetricsTTM, incomeQtr, keyMetricsAnnual, analystEst] = await Promise.all([
    fmpGet(`/profile/${ticker}`, apiKey),
    fmpGet(`/key-metrics-ttm/${ticker}`, apiKey),
    fmpGet(`/income-statement/${ticker}?period=quarter&limit=4`, apiKey),
    fmpGet(`/key-metrics/${ticker}?period=annual&limit=4`, apiKey),
    fmpGet(`/analyst-estimates/${ticker}?limit=4`, apiKey).catch(() => []),
  ]);

  if (!profile?.length) {
    console.log(`  ❌ ${ticker} — 未找到，跳过`);
    return null;
  }

  const p = profile[0];
  const km = keyMetricsTTM?.[0] ?? {};

  // ── 基础倍数 ──
  const pe = km.peRatioTTM ?? null;
  const ps = km.priceToSalesRatioTTM ?? null;

  // ── FCF Multiple = Price / FCF per share，即 pfcfRatio ──
  const fcfMultiple = km.pfcfRatioTTM ?? null;

  // ── NTM PE / NTM PS（从分析师预期计算）──
  let ntmPe = null;
  let ntmPs = null;
  let fwdFcfMultiple = null;

  if (analystEst?.length >= 4) {
    // 未来4个季度预期汇总
    const ntmRevenue = analystEst.slice(0, 4).reduce((sum, q) => sum + (q.estimatedRevenueAvg || 0), 0);
    const ntmEps = analystEst.slice(0, 4).reduce((sum, q) => sum + (q.estimatedEpsAvg || 0), 0);

    if (ntmEps > 0 && p.price) {
      ntmPe = round(p.price / ntmEps);
    }
    if (ntmRevenue > 0 && p.mktCap) {
      ntmPs = round(p.mktCap / ntmRevenue);
    }
  }

  // ── 季度毛利率 / 净利率（小数形式） ──
  // incomeQtr[0] = 最新季度 → Q4, incomeQtr[3] = 最早 → Q1
  const margins = { grossMarginQ1: null, grossMarginQ2: null, grossMarginQ3: null, grossMarginQ4: null,
                    netMarginQ1: null, netMarginQ2: null, netMarginQ3: null, netMarginQ4: null };

  if (incomeQtr?.length >= 4) {
    for (let i = 0; i < 4; i++) {
      const q = incomeQtr[i];
      const qIdx = 4 - i; // 0→Q4, 1→Q3, 2→Q2, 3→Q1
      if (q.revenue && q.revenue !== 0) {
        margins[`grossMarginQ${qIdx}`] = round(q.grossProfit / q.revenue, 4);
        margins[`netMarginQ${qIdx}`] = round(q.netIncome / q.revenue, 4);
      }
    }
  }

  // 整体毛利率/净利率取最新季度值（也可以是 TTM）
  const grossMargin = margins.grossMarginQ4;
  const netMargin = margins.netMarginQ4;

  // Non-GAAP 净利率 — FMP 无直接数据，留 null
  const nonGaapNetMargin = null;

  // ── TTM ROIC（近4年）──
  // keyMetricsAnnual[0] = 最新年 → Y4, [3] = 最早 → Y1
  const roic = { ttmRoicY1: null, ttmRoicY2: null, ttmRoicY3: null, ttmRoicY4: null };
  if (keyMetricsAnnual?.length >= 4) {
    for (let i = 0; i < 4; i++) {
      const yIdx = 4 - i; // 0→Y4, 1→Y3, 2→Y2, 3→Y1
      const val = keyMetricsAnnual[i]?.roic;
      if (val != null) roic[`ttmRoicY${yIdx}`] = round(val, 4);
    }
  } else if (keyMetricsAnnual?.length > 0) {
    // 有多少填多少
    for (let i = 0; i < keyMetricsAnnual.length; i++) {
      const yIdx = keyMetricsAnnual.length - i;
      const val = keyMetricsAnnual[i]?.roic;
      if (val != null) roic[`ttmRoicY${yIdx + (4 - keyMetricsAnnual.length)}`] = round(val, 4);
    }
  }

  const result = {
    ticker,
    companyName: p.companyName || ticker,
    snapshotDate: new Date().toISOString().slice(0, 10),
    pe: round(pe),
    ps: round(ps),
    ntmPe: round(ntmPe),
    ntmPs: round(ntmPs),
    fcfMultiple: round(fcfMultiple),
    fwdFcfMultiple: round(fwdFcfMultiple),  // 暂无数据源，留 null
    grossMargin,
    ...margins,
    netMargin,
    nonGaapNetMargin,
    ...roic,
    notes: `自动拉取 @ ${new Date().toISOString().slice(0, 10)}`,
  };

  console.log(`  ✅ ${ticker} — ${p.companyName} (PE=${fmtV(pe)}, PS=${fmtV(ps)}, ROIC=${fmtPct(roic.ttmRoicY4)})`);
  return result;
}

// ── 工具函数 ─────────────────────────────────────────────────────────────

function round(v, decimals = 2) {
  if (v == null || isNaN(v)) return null;
  const factor = 10 ** decimals;
  return Math.round(v * factor) / factor;
}

function fmtV(v) {
  return v != null ? v.toFixed(1) : 'N/A';
}

function fmtPct(v) {
  return v != null ? (v * 100).toFixed(1) + '%' : 'N/A';
}

// ── 主流程 ───────────────────────────────────────────────────────────────

async function main() {
  const { key: cliKey, tickers } = parseArgs();

  if (tickers.length === 0) {
    console.log(`
╔══════════════════════════════════════════════════════════════╗
║  估值数据拉取工具 (Financial Modeling Prep)                  ║
╚══════════════════════════════════════════════════════════════╝

用法:
  node scripts/fetch-valuations.mjs AAPL MSFT GOOGL
  node scripts/fetch-valuations.mjs AAPL --key YOUR_FMP_KEY
  node scripts/fetch-valuations.mjs --tickers AAPL,MSFT,GOOGL

API Key 配置:
  1. 命令行: --key YOUR_KEY
  2. 环境变量: set FMP_API_KEY=YOUR_KEY
  3. 项目根目录 .env 文件: FMP_API_KEY=YOUR_KEY

免费注册: https://financialmodelingprep.com/developer/docs/
`);
    process.exit(0);
  }

  const apiKey = getApiKey(cliKey);
  if (!apiKey) {
    console.error('❌ 未找到 FMP API Key。请通过 --key 参数、FMP_API_KEY 环境变量或 .env 文件配置。');
    console.error('   免费注册: https://financialmodelingprep.com/developer/docs/');
    process.exit(1);
  }

  console.log(`\n📊 开始拉取 ${tickers.length} 只股票的估值数据...\n`);

  const results = [];
  for (const ticker of tickers) {
    try {
      const data = await fetchTicker(ticker, apiKey);
      if (data) results.push(data);
    } catch (err) {
      console.log(`  ❌ ${ticker} — 拉取失败: ${err.message}`);
    }
    // 避免触发 rate limit
    if (tickers.length > 5) await new Promise(r => setTimeout(r, 300));
  }

  if (results.length === 0) {
    console.log('\n⚠️  没有成功拉取到任何数据');
    process.exit(1);
  }

  // 输出文件
  const today = new Date().toISOString().slice(0, 10);
  const outPath = resolve(ROOT, `valuation-import-${today}.json`);
  writeFileSync(outPath, JSON.stringify(results, null, 2), 'utf-8');

  console.log(`
╔══════════════════════════════════════════════════════════════╗
║  ✅ 完成！成功拉取 ${String(results.length).padStart(2)} / ${String(tickers.length).padStart(2)} 只股票                        ║
╠══════════════════════════════════════════════════════════════╣
║  输出文件: ${outPath.length > 48 ? '...' + outPath.slice(-45) : outPath.padEnd(48)}║
║                                                              ║
║  使用方式:                                                   ║
║  → 打开估值比较模块 → 导入 JSON → 粘贴文件内容即可           ║
╚══════════════════════════════════════════════════════════════╝
`);

  // 也输出摘要到终端
  console.log('拉取结果摘要:');
  console.log('─'.repeat(70));
  console.log('Ticker'.padEnd(8) + 'PE'.padStart(8) + 'PS'.padStart(8) + 'NTM PE'.padStart(9) +
    'FCF Mul'.padStart(10) + 'GM'.padStart(8) + 'NM'.padStart(8) + 'ROIC'.padStart(8));
  console.log('─'.repeat(70));
  for (const r of results) {
    console.log(
      r.ticker.padEnd(8) +
      fmtV(r.pe).padStart(8) +
      fmtV(r.ps).padStart(8) +
      fmtV(r.ntmPe).padStart(9) +
      fmtV(r.fcfMultiple).padStart(10) +
      fmtPct(r.grossMargin).padStart(8) +
      fmtPct(r.netMargin).padStart(8) +
      fmtPct(r.ttmRoicY4).padStart(8)
    );
  }
  console.log('─'.repeat(70));
}

main().catch(err => {
  console.error('致命错误:', err.message);
  process.exit(1);
});

File: scripts/fetch_valuations.py
#!/usr/bin/env python3
"""
估值数据拉取脚本 (Python版)

从 stockanalysis.com 爬取美股估值数据：
  - 季度毛利率/净利率 → /financials/?p=quarterly
  - 4年历史ROIC       → /financials/ratios/
  - EV/FCF TTM        → /statistics/

输出符合应用导入格式的 JSON 文件。

用法:
  python scripts/fetch_valuations.py AAPL MSFT GOOGL NVDA META AMZN NOW TSLA
  python scripts/fetch_valuations.py AAPL --output my-output.json

依赖:
  pip install -r scripts/requirements.txt

输出:
  ./valuation-import-YYYY-MM-DD.json
"""

import sys
import json
import time
import re
from datetime import date
from pathlib import Path

import requests
from bs4 import BeautifulSoup

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36"
}


# ── HTTP 请求 ──────────────────────────────────────────────────────────────

def fetch_page(url: str, retries: int = 2) -> str:
    """Fetch a page with retries and rate limit handling."""
    for attempt in range(retries + 1):
        try:
            resp = requests.get(url, headers=HEADERS, timeout=20)
            if resp.status_code == 200:
                return resp.text
            elif resp.status_code == 429:
                wait = 5 * (attempt + 1)
                print(f"    ⏸️  Rate limited, waiting {wait}s...")
                time.sleep(wait)
            else:
                print(f"    ❌ HTTP {resp.status_code}: {url}")
                return ""
        except requests.RequestException as e:
            if attempt < retries:
                time.sleep(3)
            else:
                print(f"    ❌ Request failed: {e}")
                return ""
    return ""


# ── stockanalysis: 季度毛利率 / 净利率 ────────────────────────────────────

def fetch_quarterly_margins(ticker: str) -> dict:
    """
    从 stockanalysis.com 季度利润表获取最近4个季度的毛利率和净利率。
    返回 Q1(最早) 到 Q4(最近) + TTM加权值。
    """
    result = {
        "grossMargin": None,
        "grossMarginQ1": None, "grossMarginQ2": None,
        "grossMarginQ3": None, "grossMarginQ4": None,
        "netMargin": None,
        "netMarginQ1": None, "netMarginQ2": None,
        "netMarginQ3": None, "netMarginQ4": None,
    }

    url = f"https://stockanalysis.com/stocks/{ticker.lower()}/financials/?p=quarterly"
    html = fetch_page(url)
    if not html:
        return result

    soup = BeautifulSoup(html, "html.parser")
    table = soup.find("table")
    if not table:
        print(f"    ⚠️  {ticker}: 季度财务表格未找到")
        return result

    # Parse table into a dict of {row_label: [values]}
    rows = table.find_all("tr")
    data = {}
    num_cols = 0
    for row in rows:
        cells = row.find_all(["td", "th"])
        if len(cells) >= 2:
            label = cells[0].get_text(strip=True)
            values = [c.get_text(strip=True) for c in cells[1:]]
            data[label] = values
            num_cols = max(num_cols, len(values))

    # Extract Revenue, Gross Profit, Net Income for the 4 most recent quarters
    # Columns are in reverse chronological order (newest first)
    revenue_row = data.get("Revenue", [])
    gp_row = data.get("Gross Profit", [])
    ni_row = data.get("Net Income", [])

    if not revenue_row:
        print(f"    ⚠️  {ticker}: Revenue行未找到")
        return result

    # Take first 4 columns (most recent 4 quarters), then reverse to get Q1=oldest
    n = min(4, len(revenue_row))
    revenues = [_parse_number(revenue_row[i]) for i in range(n)]
    gross_profits = [_parse_number(gp_row[i]) if i < len(gp_row) else None for i in range(n)]
    net_incomes = [_parse_number(ni_row[i]) if i < len(ni_row) else None for i in range(n)]

    # Reverse: index 0 becomes Q4 (most recent) → after reverse, index 0 = oldest = Q1
    revenues.reverse()
    gross_profits.reverse()
    net_incomes.reverse()

    total_rev = 0.0
    total_gp = 0.0
    total_ni = 0.0

    for i in range(len(revenues)):
        rev = revenues[i]
        gp = gross_profits[i]
        ni = net_incomes[i]

        if rev and rev != 0:
            gm = round(gp / rev, 4) if gp is not None else None
            nm = round(ni / rev, 4) if ni is not None else None
        else:
            gm, nm = None, None

        q_num = i + 1  # 1=oldest (Q1), 4=newest (Q4)
        result[f"grossMarginQ{q_num}"] = gm
        result[f"netMarginQ{q_num}"] = nm

        if rev:
            total_rev += rev
        if gp:
            total_gp += gp
        if ni:
            total_ni += ni

    # TTM weighted margins
    if total_rev > 0:
        result["grossMargin"] = round(total_gp / total_rev, 4) if total_gp else None
        result["netMargin"] = round(total_ni / total_rev, 4) if total_ni else None

    return result


# ── stockanalysis: EV/FCF ──────────────────────────────────────────────────

def fetch_ev_fcf(ticker: str) -> dict:
    """
    从 stockanalysis.com statistics 页面获取 EV 和 FCF，计算 EV/FCF。
    """
    result = {"fcfMultiple": None}

    url = f"https://stockanalysis.com/stocks/{ticker.lower()}/statistics/"
    html = fetch_page(url)
    if not html:
        return result

    soup = BeautifulSoup(html, "html.parser")

    ev = None
    fcf = None

    # Search all tables for Enterprise Value and Free Cash Flow
    for table in soup.find_all("table"):
        for row in table.find_all("tr"):
            cells = row.find_all(["td", "th"])
            if len(cells) >= 2:
                label = cells[0].get_text(strip=True)
                value = cells[1].get_text(strip=True)

                if "Enterprise Value" in label and "EV/" not in label:
                    ev = _parse_large_number(value)
                elif label == "Free Cash Flow":
                    fcf = _parse_large_number(value)

    if ev and fcf and fcf > 0:
        result["fcfMultiple"] = round(ev / fcf, 2)
    elif fcf is not None and fcf <= 0:
        print(f"    ⚠️  {ticker}: FCF为负 ({fcf:,.0f}), fcfMultiple设为null")

    return result


# ── stockanalysis: 4年 ROIC 历史 ──────────────────────────────────────────

def fetch_roic_history(ticker: str) -> dict:
    """
    从 stockanalysis.com 年度 ratios 页面获取近4年的 ROIC。
    返回 Y1(最早) 到 Y4(最近)。
    """
    result = {"ttmRoicY1": None, "ttmRoicY2": None, "ttmRoicY3": None, "ttmRoicY4": None}

    url = f"https://stockanalysis.com/stocks/{ticker.lower()}/financials/ratios/"
    html = fetch_page(url)
    if not html:
        return result

    soup = BeautifulSoup(html, "html.parser")
    table = soup.find("table")
    if not table:
        print(f"    ⚠️  {ticker}: ratios表格未找到")
        return result

    # Find the ROIC row
    for row in table.find_all("tr"):
        cells = row.find_all(["td", "th"])
        if cells and "ROIC" in cells[0].get_text():
            # Values: [label, Current, FY_newest, FY_next, ..., FY_oldest]
            # Skip "Current" (index 1), take next 4 FY values (index 2-5)
            values = [c.get_text(strip=True) for c in cells[1:]]

            # First value is "Current" (TTM), skip it; take next 4 which are fiscal years
            fy_values = []
            for v in values[1:]:  # Skip "Current"
                pct = _parse_pct(v)
                if pct is not None:
                    fy_values.append(pct)
                if len(fy_values) >= 4:
                    break

            # fy_values[0] = most recent FY, fy_values[3] = oldest FY
            # Reverse to get Y1=oldest, Y4=newest
            fy_values.reverse()
            for i, val in enumerate(fy_values):
                result[f"ttmRoicY{i + 1}"] = round(val, 4)

            break

    return result


# ── 数值解析工具函数 ─────────────────────────────────────────────────────────

def _parse_number(text: str):
    """解析财务数字 (e.g. '111,184' or '-5,234' or '94.03B')"""
    if not text or text == "-" or text == "n/a":
        return None
    text = text.strip().replace(",", "")
    try:
        return float(text)
    except ValueError:
        return _parse_large_number(text)


def _parse_large_number(text: str):
    """解析带单位的数字 (e.g. '4.35T', '129.17B', '523.4M')"""
    if not text or text == "-" or text == "n/a":
        return None
    text = text.strip().replace(",", "").replace("$", "")

    multipliers = {"T": 1e12, "B": 1e9, "M": 1e6, "K": 1e3}
    for suffix, mult in multipliers.items():
        if text.endswith(suffix):
            try:
                return float(text[:-1]) * mult
            except ValueError:
                return None
    try:
        return float(text)
    except ValueError:
        return None


def _parse_pct(text: str):
    """解析百分比字符串, 返回小数 (e.g. '48.24%' → 0.4824)"""
    if not text or text == "-" or text == "n/a":
        return None
    text = text.strip().replace(",", "").replace("%", "")
    try:
        return float(text) / 100.0
    except ValueError:
        return None


# ── 组装单只股票数据 ────────────────────────────────────────────────────────

def build_valuation(ticker: str) -> dict:
    """获取单只股票的所有估值数据"""
    print(f"  ⏳ {ticker} — 拉取中...")

    # 1. 季度利润率
    margins = fetch_quarterly_margins(ticker)
    time.sleep(1.5)

    # 2. EV/FCF
    ev_fcf = fetch_ev_fcf(ticker)
    time.sleep(1.5)

    # 3. 历史 ROIC
    roic = fetch_roic_history(ticker)
    time.sleep(1.5)

    # 4. 公司名称 (从 statistics 页面的 title 标签获取)
    company_name = ticker  # fallback

    # 组装
    result = {
        "ticker": ticker,
        "companyName": company_name,
        "pe": None,
        "ps": None,
        "ntmPe": None,
        "ntmPs": None,
        "grossMargin": margins["grossMargin"],
        "grossMarginQ1": margins["grossMarginQ1"],
        "grossMarginQ2": margins["grossMarginQ2"],
        "grossMarginQ3": margins["grossMarginQ3"],
        "grossMarginQ4": margins["grossMarginQ4"],
        "netMargin": margins["netMargin"],
        "nonGaapNetMargin": None,
        "netMarginQ1": margins["netMarginQ1"],
        "netMarginQ2": margins["netMarginQ2"],
        "netMarginQ3": margins["netMarginQ3"],
        "netMarginQ4": margins["netMarginQ4"],
        "fcfMultiple": ev_fcf["fcfMultiple"],
        "fwdFcfMultiple": None,
        "ttmRoicY1": roic["ttmRoicY1"],
        "ttmRoicY2": roic["ttmRoicY2"],
        "ttmRoicY3": roic["ttmRoicY3"],
        "ttmRoicY4": roic["ttmRoicY4"],
        "notes": f"自动拉取 @ {date.today().isoformat()} | stockanalysis.com",
    }

    # 打印摘要
    gm = margins["grossMarginQ4"]
    nm = margins["netMarginQ4"]
    fcf = ev_fcf["fcfMultiple"]
    r4 = roic["ttmRoicY4"]
    print(f"  ✅ {ticker}")
    print(f"     GM={_fmt_pct(gm)} NM={_fmt_pct(nm)} EV/FCF={_fmt_val(fcf)} ROIC(Y4)={_fmt_pct(r4)}")

    return result


# ── 格式化工具函数 ────────────────────────────────────────────────────────

def _fmt_pct(v):
    return f"{v*100:.1f}%" if v is not None else "N/A"

def _fmt_val(v):
    return f"{v:.1f}" if v is not None else "N/A"


# ── CLI 参数解析 ───────────────────────────────────────────────────────────

def parse_args():
    args = sys.argv[1:]
    tickers = []
    output = None

    i = 0
    while i < len(args):
        if args[i] == "--output" and i + 1 < len(args):
            output = args[i + 1]
            i += 2
        elif args[i] == "--tickers" and i + 1 < len(args):
            tickers.extend(t.strip().upper() for t in args[i + 1].split(","))
            i += 2
        elif not args[i].startswith("--"):
            tickers.append(args[i].upper())
            i += 1
        else:
            i += 1

    return tickers, output


# ── 主流程 ─────────────────────────────────────────────────────────────────

def main():
    tickers, output_path = parse_args()

    if not tickers:
        print("""
╔══════════════════════════════════════════════════════════════╗
║  估值数据拉取工具 (stockanalysis.com)                        ║
╚══════════════════════════════════════════════════════════════╝

用法:
  python scripts/fetch_valuations.py AAPL MSFT GOOGL NVDA
  python scripts/fetch_valuations.py --tickers AAPL,MSFT,GOOGL
  python scripts/fetch_valuations.py AAPL --output my-file.json

数据来源 (全部来自 stockanalysis.com):
  • 季度毛利率/净利率 → /financials/?p=quarterly
  • EV/FCF TTM        → /statistics/
  • 历史ROIC (4年)    → /financials/ratios/

安装依赖:
  pip install -r scripts/requirements.txt
""")
        sys.exit(0)

    print(f"\n📊 开始拉取 {len(tickers)} 只股票的估值数据...\n")

    results = []
    for ticker in tickers:
        try:
            data = build_valuation(ticker)
            results.append(data)
        except Exception as e:
            print(f"  ❌ {ticker} — 拉取失败: {e}")
        print()

    if not results:
        print("⚠️  没有成功拉取到任何数据")
        sys.exit(1)

    # 输出文件
    if not output_path:
        root = Path(__file__).resolve().parent.parent
        output_path = root / f"valuation-import-{date.today().isoformat()}.json"
    else:
        output_path = Path(output_path)

    output_path.write_text(json.dumps(results, indent=2, ensure_ascii=False), encoding="utf-8")

    # 打印摘要
    print("═" * 72)
    print(f"  ✅ 完成！成功拉取 {len(results)} / {len(tickers)} 只股票")
    print(f"  📁 输出文件: {output_path}")
    print("═" * 72)
    print()
    print(f"{'Ticker':<8}{'GM':>8}{'NM':>8}{'EV/FCF':>10}{'ROIC(Y4)':>10}")
    print("─" * 44)
    for r in results:
        print(
            f"{r['ticker']:<8}"
            f"{_fmt_pct(r['grossMarginQ4']):>8}"
            f"{_fmt_pct(r['netMarginQ4']):>8}"
            f"{_fmt_val(r['fcfMultiple']):>10}"
            f"{_fmt_pct(r['ttmRoicY4']):>10}"
        )
    print("─" * 44)


if __name__ == "__main__":
    main()
